Exceptions in throttled tasks are propagated to client at wrong time

Consider this scenario:
1. I create a file with language not supported by the Annotace service
2. I attempt to run text analysis once. It fails, the log contains the following message:

```
18-11-2024 16:17:54.170 [] [TermItScheduler-3] TRACE c.c.k.t.u.t.ThrottleAspect - Running throttled task [0 left] [1 running] 'ThrottleAspect.Identifier{group='',identifier='TextAnalysisService.analyzeFile(..)-http://onto.fel.cvut.cz/ontologies/slovnik/ml-test/document/soubor/uap.html'}'
18-11-2024 16:17:55.879 [] [TermItScheduler-3] TRACE c.c.k.t.u.l.LongRunningTasksRegistry - Long running task changed state: fileAnalysis{RUNNING, startedAt=2024-11-18T15:17:55Z, 67a48b80-8a67-4b95-890b-acbb30f991a9}
18-11-2024 16:17:55.880 [] [TermItScheduler-3] DEBUG c.c.k.t.s.d.DefaultDocumentManager - Loading file content from /home/kidney/Development/Java/Projects/termit/data/dokument-pro-ml-test_-1891243467/uap.html.
18-11-2024 16:17:55.882 [halsey@unsc.org] [http-nio-8080-exec-3] TRACE c.c.k.t.u.l.LongRunningTasksRegistry - Long running task changed state: fileAnalysis{RUNNING, startedAt=2024-11-18T15:17:55Z, 67a48b80-8a67-4b95-890b-acbb30f991a9}
18-11-2024 16:17:55.883 [] [TermItScheduler-3] DEBUG c.c.k.t.s.d.TextAnalysisService - Invoking text analysis service at 'http://localhost:8081/annotace/annotate' on input: TextAnalysisInput{content='<!DOCTYPE html>
<html>
<head>
	<meta http-equiv="c...', vocabularyRepository=http://localhost:7200/repositories/termit, vocabularyContexts=[http://onto.fel.cvut.cz/ontologies/slovnik/ml-test], language=de}
18-11-2024 16:17:55.935 [] [TermItScheduler-3] ERROR c.c.k.t.u.t.ThrottleAspect - Exception thrown during task execution
cz.cvut.kbss.termit.exception.UnsupportedTextAnalysisLanguageException: No tagger for language 'de' available.
	at cz.cvut.kbss.termit.service.document.TextAnalysisService.handleTextAnalysisInvocationClientException(TextAnalysisService.java:178)
	at cz.cvut.kbss.termit.service.document.TextAnalysisService.invokeTextAnalysisOnFile(TextAnalysisService.java:136)
	at cz.cvut.kbss.termit.service.document.TextAnalysisService.analyzeFile_aroundBody0(TextAnalysisService.java:103)
	at cz.cvut.kbss.termit.service.document.TextAnalysisService$AjcClosure1.run(TextAnalysisService.java:1)
	at org.springframework.transaction.aspectj.AbstractTransactionAspect.ajc$around$org_springframework_transaction_aspectj_AbstractTransactionAspect$1$2a73e96cproceed(AbstractTransactionAspect.aj:67)
	at org.springframework.transaction.aspectj.AbstractTransactionAspect$AbstractTransactionAspect$1.proceedWithInvocation(AbstractTransactionAspect.aj:73)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:379)
	at org.springframework.transaction.aspectj.AbstractTransactionAspect.ajc$around$org_springframework_transaction_aspectj_AbstractTransactionAspect$1$2a73e96c(AbstractTransactionAspect.aj:71)
	at cz.cvut.kbss.termit.service.document.TextAnalysisService.analyzeFile(TextAnalysisService.java:100)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:355)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:768)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:379)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:768)
	at org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:89)
	at cz.cvut.kbss.termit.util.throttle.ThrottleAspect.lambda$4(ThrottleAspect.java:345)
	at cz.cvut.kbss.termit.util.throttle.ThrottledFuture.run(ThrottledFuture.java:203)
	at cz.cvut.kbss.termit.util.throttle.ThrottleAspect.lambda$7(ThrottleAspect.java:405)
	at cz.cvut.kbss.termit.util.throttle.SynchronousTransactionExecutor.execute_aroundBody0(SynchronousTransactionExecutor.java:20)
	at cz.cvut.kbss.termit.util.throttle.SynchronousTransactionExecutor$AjcClosure1.run(SynchronousTransactionExecutor.java:1)
	at org.springframework.transaction.aspectj.AbstractTransactionAspect.ajc$around$org_springframework_transaction_aspectj_AbstractTransactionAspect$1$2a73e96cproceed(AbstractTransactionAspect.aj:67)
	at org.springframework.transaction.aspectj.AbstractTransactionAspect$AbstractTransactionAspect$1.proceedWithInvocation(AbstractTransactionAspect.aj:73)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:379)
	at org.springframework.transaction.aspectj.AbstractTransactionAspect.ajc$around$org_springframework_transaction_aspectj_AbstractTransactionAspect$1$2a73e96c(AbstractTransactionAspect.aj:71)
	at cz.cvut.kbss.termit.util.throttle.SynchronousTransactionExecutor.execute(SynchronousTransactionExecutor.java:19)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:355)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:768)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:379)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:768)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:720)
	at cz.cvut.kbss.termit.util.throttle.SynchronousTransactionExecutor$$SpringCGLIB$$0.execute(<generated>)
	at cz.cvut.kbss.termit.util.throttle.ThrottleAspect.lambda$6(ThrottleAspect.java:405)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
18-11-2024 16:18:06.293 [] [TermItScheduler-3] TRACE c.c.k.t.u.l.LongRunningTasksRegistry - Long running task changed state: fileAnalysis{DONE, startedAt=2024-11-18T15:17:55Z, 67a48b80-8a67-4b95-890b-acbb30f991a9}
18-11-2024 16:18:06.295 [] [TermItScheduler-3] TRACE c.c.k.t.u.l.LongRunningTasksRegistry - All long running tasks completed
18-11-2024 16:18:06.301 [] [TermItScheduler-3] TRACE c.c.k.t.u.t.ThrottleAspect - Finished throttled task [0 left] [0 running] 'ThrottleAspect.Identifier{group='',identifier='TextAnalysisService.analyzeFile(..)-http://onto.fel.cvut.cz/ontologies/slovnik/ml-test/document/soubor/uap.html'}'
18-11-2024 16:19:12.247 [] [TermItScheduler-3] ERROR o.s.s.s.TaskUtils$LoggingErrorHandler - Unexpected error occurred in scheduled task
org.springframework.transaction.TransactionSystemException: Unable to commit JOPA entity transaction!
	at com.github.ledsoft.jopa.spring.transaction.JopaTransactionManager.doCommit(JopaTransactionManager.java:83)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:795)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:758)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.commitTransactionAfterReturning(TransactionAspectSupport.java:663)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:413)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:768)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:720)
	at cz.cvut.kbss.termit.util.throttle.SynchronousTransactionExecutor$$SpringCGLIB$$0.execute(<generated>)
	at cz.cvut.kbss.termit.util.throttle.ThrottleAspect.lambda$6(ThrottleAspect.java:405)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: cz.cvut.kbss.jopa.exceptions.RollbackException: Trying to commit transaction marked as rollback only.
	at cz.cvut.kbss.jopa.transactions.EntityTransactionImpl.commit(EntityTransactionImpl.java:56)
	at com.github.ledsoft.jopa.spring.transaction.JopaTransactionManager.doCommit(JopaTransactionManager.java:80)
	... 18 common frames omitted
```
The second exception (the one about attempting to commit a rollback-only transaction) is strange, the transaction should have finished with the first exception. But this is not the biggest issue.

3. I attempt text analysis again. This time, the call immediately fails (500 returned by backend API) with an error about unsupported language, but the error is apparently from the previously throttled task, because it is thrown synchronously, when attempting to schedule the new analysis task.
![image](https://github.com/user-attachments/assets/4dc07891-30f4-4cb2-99ab-e6f26ca5d6af)

I was able to trace the synchronous error to the following call in ThrottleAspect#L269.
```java
oldThrottledFuture.then(ignored -> schedule(identifier, pair.getFirst(),
                                                            throttleExpired && oldFutureIsDone)
                );
```
The `oldThrottledFuture` finished with an exception (step 2 above), but its resolution is still attempted.

--- a/src/main/java/cz/cvut/kbss/termit/persistence/dao/VocabularyDao.java
+++ b/src/main/java/cz/cvut/kbss/termit/persistence/dao/VocabularyDao.java
@@ -48,7 +48,7 @@ import cz.cvut.kbss.termit.persistence.validation.VocabularyContentValidator;
 import cz.cvut.kbss.termit.service.snapshot.SnapshotProvider;
 import cz.cvut.kbss.termit.util.Configuration;
 import cz.cvut.kbss.termit.util.Utils;
-import cz.cvut.kbss.termit.util.throttle.CacheableFuture;
+import cz.cvut.kbss.termit.util.throttle.ThrottledFuture;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
@@ -366,7 +366,7 @@ public class VocabularyDao extends BaseAssetDao<Vocabulary>
     }
 
     @Transactional
-    public CacheableFuture<Collection<ValidationResult>> validateContents(URI vocabulary) {
+    public ThrottledFuture<Collection<ValidationResult>> validateContents(URI vocabulary) {
         final VocabularyContentValidator validator = context.getBean(VocabularyContentValidator.class);
         final Collection<URI> importClosure = getTransitivelyImportedVocabularies(vocabulary);
         importClosure.add(vocabulary);

--- a/src/main/java/cz/cvut/kbss/termit/service/business/VocabularyService.java
+++ b/src/main/java/cz/cvut/kbss/termit/service/business/VocabularyService.java
@@ -46,8 +46,8 @@ import cz.cvut.kbss.termit.util.Configuration;
 import cz.cvut.kbss.termit.util.TypeAwareClasspathResource;
 import cz.cvut.kbss.termit.util.TypeAwareFileSystemResource;
 import cz.cvut.kbss.termit.util.TypeAwareResource;
-import cz.cvut.kbss.termit.util.throttle.CacheableFuture;
 import cz.cvut.kbss.termit.util.throttle.Throttle;
+import cz.cvut.kbss.termit.util.throttle.ThrottledFuture;
 import jakarta.annotation.Nonnull;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -384,7 +384,7 @@ public class VocabularyService
      *
      * @param vocabulary Vocabulary to validate
      */
-    public CacheableFuture<Collection<ValidationResult>> validateContents(URI vocabulary) {
+    public ThrottledFuture<Collection<ValidationResult>> validateContents(URI vocabulary) {
         return repositoryService.validateContents(vocabulary);
     }
 

--- a/src/main/java/cz/cvut/kbss/termit/service/repository/VocabularyRepositoryService.java
+++ b/src/main/java/cz/cvut/kbss/termit/service/repository/VocabularyRepositoryService.java
@@ -42,7 +42,7 @@ import cz.cvut.kbss.termit.service.snapshot.SnapshotProvider;
 import cz.cvut.kbss.termit.util.Configuration;
 import cz.cvut.kbss.termit.util.Constants;
 import cz.cvut.kbss.termit.util.Utils;
-import cz.cvut.kbss.termit.util.throttle.CacheableFuture;
+import cz.cvut.kbss.termit.util.throttle.ThrottledFuture;
 import cz.cvut.kbss.termit.workspace.EditableVocabularies;
 import jakarta.annotation.Nonnull;
 import jakarta.validation.Validator;
@@ -334,7 +334,7 @@ public class VocabularyRepositoryService extends BaseAssetRepositoryService<Voca
         }
     }
 
-    public CacheableFuture<Collection<ValidationResult>> validateContents(URI vocabulary) {
+    public ThrottledFuture<Collection<ValidationResult>> validateContents(URI vocabulary) {
         return vocabularyDao.validateContents(vocabulary);
     }
 

--- a/src/main/java/cz/cvut/kbss/termit/util/throttle/CacheableFuture.java
+++ b/src/main/java/cz/cvut/kbss/termit/util/throttle/CacheableFuture.java
@@ -11,7 +11,7 @@ import java.util.concurrent.Future;
  * A future which can provide a cached result before its completion.
  * @see Future
  */
-public interface CacheableFuture<T> extends ChainableFuture<T> {
+public interface CacheableFuture<T> extends Future<T> {
 
     /**
      * @return the cached result when available

--- a/src/main/java/cz/cvut/kbss/termit/util/throttle/ChainableFuture.java
+++ b/src/main/java/cz/cvut/kbss/termit/util/throttle/ChainableFuture.java
@@ -3,14 +3,15 @@ package cz.cvut.kbss.termit.util.throttle;
 import java.util.concurrent.Future;
 import java.util.function.Consumer;
 
-public interface ChainableFuture<T> extends Future<T> {
+public interface ChainableFuture<T, F extends ChainableFuture<T, F>> extends Future<T> {
 
     /**
-     * Executes this action once the future is completed normally.
-     * Action is not executed on exceptional completion.
+     * Executes this action once the future is completed.
+     * Action is executed no matter if the future is completed successfully, exceptionally or cancelled.
      * <p>
-     * If the future is already completed, action is executed synchronously.
-     * @param action action to be executed
+     * If the future is already completed, it is executed synchronously.
+     * @param action action receiving this future after completion
+     * @return this future
      */
-    ChainableFuture then(Consumer<T> action);
+    ChainableFuture<T, F> then(Consumer<F> action);
 }

--- a/src/main/java/cz/cvut/kbss/termit/util/throttle/ChainableFuture.java
+++ b/src/main/java/cz/cvut/kbss/termit/util/throttle/ChainableFuture.java
@@ -10,6 +10,8 @@ public interface ChainableFuture<T, F extends ChainableFuture<T, F>> extends Fut
      * Action is executed no matter if the future is completed successfully, exceptionally or cancelled.
      * <p>
      * If the future is already completed, it is executed synchronously.
+     * <p>
+     * Note that you <b>must</b> use the future passed as the parameter and not the original future object.
      * @param action action receiving this future after completion
      * @return this future
      */

--- a/src/main/java/cz/cvut/kbss/termit/util/throttle/ThrottledFuture.java
+++ b/src/main/java/cz/cvut/kbss/termit/util/throttle/ThrottledFuture.java
@@ -245,7 +245,7 @@ public class ThrottledFuture<T> implements CacheableFuture<T>, LongRunningTask {
     public ThrottledFuture<T> then(Consumer<T> action) {
         try {
             callbackLock.lock();
-            if (future.isDone() && !future.isCancelled()) {
+            if (future.isDone() && !future.isCancelled() && !future.isCompletedExceptionally()) {
                 try {
                     action.accept(future.get());
                 } catch (InterruptedException e) {
@@ -261,5 +261,13 @@ public class ThrottledFuture<T> implements CacheableFuture<T>, LongRunningTask {
             callbackLock.unlock();
         }
         return this;
+    }
+
+    /**
+     * @return {@code true} if this future completed
+     * exceptionally
+     */
+    public boolean isCompletedExceptionally() {
+        return future.isCompletedExceptionally();
     }
 }

--- a/src/main/java/cz/cvut/kbss/termit/util/throttle/ThrottledFuture.java
+++ b/src/main/java/cz/cvut/kbss/termit/util/throttle/ThrottledFuture.java
@@ -1,6 +1,5 @@
 package cz.cvut.kbss.termit.util.throttle;
 
-import cz.cvut.kbss.termit.exception.TermItException;
 import cz.cvut.kbss.termit.util.Utils;
 import cz.cvut.kbss.termit.util.longrunning.LongRunningTask;
 import jakarta.annotation.Nonnull;
@@ -20,7 +19,7 @@ import java.util.concurrent.locks.ReentrantLock;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
 
-public class ThrottledFuture<T> implements CacheableFuture<T>, LongRunningTask {
+public class ThrottledFuture<T> implements CacheableFuture<T>, ChainableFuture<T, ThrottledFuture<T>>, LongRunningTask {
 
     private final ReentrantLock lock = new ReentrantLock();
     private final ReentrantLock callbackLock = new ReentrantLock();
@@ -33,7 +32,7 @@ public class ThrottledFuture<T> implements CacheableFuture<T>, LongRunningTask {
 
     private @Nullable Supplier<T> task;
 
-    private final List<Consumer<T>> onCompletion = new ArrayList<>();
+    private final List<Consumer<ThrottledFuture<T>>> onCompletion = new ArrayList<>();
 
     private final AtomicReference<Instant> startedAt = new AtomicReference<>(null);
 
@@ -90,7 +89,16 @@ public class ThrottledFuture<T> implements CacheableFuture<T>, LongRunningTask {
 
     @Override
     public boolean cancel(boolean mayInterruptIfRunning) {
-        return future.cancel(mayInterruptIfRunning);
+        if(!future.cancel(mayInterruptIfRunning)) {
+            return false;
+        }
+
+        if (task != null) {
+            callbackLock.lock();
+            onCompletion.forEach(c -> c.accept(this));
+            callbackLock.unlock();
+        }
+        return true;
     }
 
     @Override
@@ -124,7 +132,7 @@ public class ThrottledFuture<T> implements CacheableFuture<T>, LongRunningTask {
      * @return If the current task is already running, was canceled or already completed, returns a new future for the given task.
      * Otherwise, replaces the current task and returns self.
      */
-    protected ThrottledFuture<T> update(Supplier<T> task, @Nonnull List<Consumer<T>> onCompletion) {
+    protected ThrottledFuture<T> update(Supplier<T> task, @Nonnull List<Consumer<ThrottledFuture<T>>> onCompletion) {
         boolean locked = false;
         try {
             locked = lock.tryLock();
@@ -201,14 +209,16 @@ public class ThrottledFuture<T> implements CacheableFuture<T>, LongRunningTask {
                 T result = null;
                 if (task != null) {
                     result = task.get();
-                    final T finalResult = result;
-                    callbackLock.lock();
-                    onCompletion.forEach(c -> c.accept(finalResult));
-                    callbackLock.unlock();
                 }
                 future.complete(result);
             } catch (Exception e) {
                 future.completeExceptionally(e);
+            } finally {
+                if (task != null) {
+                    callbackLock.lock();
+                    onCompletion.forEach(c -> c.accept(this));
+                    callbackLock.unlock();
+                }
             }
         } finally {
             if (locked) {
@@ -242,18 +252,11 @@ public class ThrottledFuture<T> implements CacheableFuture<T>, LongRunningTask {
     }
 
     @Override
-    public ThrottledFuture<T> then(Consumer<T> action) {
+    public ThrottledFuture<T> then(Consumer<ThrottledFuture<T>> action) {
         try {
             callbackLock.lock();
-            if (future.isDone() && !future.isCancelled() && !future.isCompletedExceptionally()) {
-                try {
-                    action.accept(future.get());
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                    throw new TermItException(e);
-                } catch (ExecutionException e) {
-                    throw new TermItException(e);
-                }
+            if (future.isDone()) {
+                action.accept(this);
             } else {
                 onCompletion.add(action);
             }

--- a/src/main/java/cz/cvut/kbss/termit/util/throttle/ThrottledFuture.java
+++ b/src/main/java/cz/cvut/kbss/termit/util/throttle/ThrottledFuture.java
@@ -89,11 +89,12 @@ public class ThrottledFuture<T> implements CacheableFuture<T>, ChainableFuture<T
 
     @Override
     public boolean cancel(boolean mayInterruptIfRunning) {
+        final boolean wasCanceled = isCancelled();
         if(!future.cancel(mayInterruptIfRunning)) {
             return false;
         }
 
-        if (task != null) {
+        if (!wasCanceled && task != null) {
             callbackLock.lock();
             onCompletion.forEach(c -> c.accept(this));
             callbackLock.unlock();
@@ -268,7 +269,7 @@ public class ThrottledFuture<T> implements CacheableFuture<T>, ChainableFuture<T
 
     /**
      * @return {@code true} if this future completed
-     * exceptionally
+     * exceptionally or was cancelled.
      */
     public boolean isCompletedExceptionally() {
         return future.isCompletedExceptionally();

--- a/src/main/java/cz/cvut/kbss/termit/websocket/VocabularySocketController.java
+++ b/src/main/java/cz/cvut/kbss/termit/websocket/VocabularySocketController.java
@@ -11,7 +11,7 @@ import cz.cvut.kbss.termit.service.IdentifierResolver;
 import cz.cvut.kbss.termit.service.business.VocabularyService;
 import cz.cvut.kbss.termit.util.Configuration;
 import cz.cvut.kbss.termit.util.Constants;
-import cz.cvut.kbss.termit.util.throttle.CacheableFuture;
+import cz.cvut.kbss.termit.util.throttle.ThrottledFuture;
 import jakarta.annotation.Nonnull;
 import org.springframework.context.event.EventListener;
 import org.springframework.messaging.MessageHeaders;
@@ -53,7 +53,7 @@ public class VocabularySocketController extends BaseWebSocketController {
         final URI identifier = resolveIdentifier(namespace.orElse(config.getNamespace().getVocabulary()), localName);
         final Vocabulary vocabulary = vocabularyService.getReference(identifier);
 
-        final CacheableFuture<Collection<ValidationResult>> future = vocabularyService.validateContents(vocabulary.getUri());
+        final ThrottledFuture<Collection<ValidationResult>> future = vocabularyService.validateContents(vocabulary.getUri());
 
         future.getNow().ifPresentOrElse(validationResults ->
             // if there is a result present (returned from cache), send it
@@ -66,14 +66,15 @@ public class VocabularySocketController extends BaseWebSocketController {
                     messageHeaders
             ), () ->
             // otherwise reply will be sent once the future is resolved
-            future.then(results ->
+            future.then(completedFuture ->
+                    completedFuture.getNow().ifPresent(results ->
                 sendToSession(
                         WebSocketDestinations.VOCABULARIES_VALIDATION,
                         results,
                         getHeaders(identifier,
                                 Map.of("cached", false)),
                         messageHeaders
-                ))
+                )))
         );
 
     }

--- a/src/test/java/cz/cvut/kbss/termit/util/throttle/ThrottledFutureTest.java
+++ b/src/test/java/cz/cvut/kbss/termit/util/throttle/ThrottledFutureTest.java
@@ -139,6 +139,18 @@ class ThrottledFutureTest {
     }
 
     @Test
+    void thenActionIsNotExecutedWhenFutureCompletedExceptionally() {
+        final AtomicBoolean completed = new AtomicBoolean(false);
+        final ThrottledFuture<?> future = ThrottledFuture.of(() -> {
+            throw new RuntimeException();
+        });
+        future.run(null);
+        assertFalse(completed.get());
+        future.then(futureResult -> completed.set(true));
+        assertFalse(completed.get());
+    }
+
+    @Test
     void callingRunWillExecuteFutureOnlyOnce() {
         AtomicInteger count = new AtomicInteger(0);
         final ThrottledFuture<?> future = ThrottledFuture.of(() -> {

--- a/src/test/java/cz/cvut/kbss/termit/util/throttle/ThrottledFutureTest.java
+++ b/src/test/java/cz/cvut/kbss/termit/util/throttle/ThrottledFutureTest.java
@@ -137,7 +137,59 @@ class ThrottledFutureTest {
     }
 
     @Test
-    void thenActionIsExecutedWhenFutureCompletedExceptionally() {
+    void thenActionIsExecutedOnlyOnceWhenFutureIsCancelled() {
+        final AtomicInteger executionCount = new AtomicInteger(0);
+        final ThrottledFuture<?> future = ThrottledFuture.of(() -> null);
+        future.then(f -> executionCount.incrementAndGet());
+        assertEquals(0, executionCount.get());
+        future.cancel(false);
+        assertEquals(1, executionCount.get());
+        future.cancel(false);
+        future.cancel(true);
+        assertEquals(1, executionCount.get());
+    }
+
+    @Test
+    void thenActionIsExecutedWhenFutureCompletesExceptionally() {
+        final AtomicBoolean completed = new AtomicBoolean(false);
+        final ThrottledFuture<?> future = ThrottledFuture.of(() -> {
+            throw new RuntimeException();
+        });
+        future.then(futureResult -> completed.set(true));
+        assertFalse(completed.get());
+        future.run(null);
+        assertTrue(completed.get());
+    }
+
+    @Test
+    void isCompletedExceptionallyReturnsTrueWhenFutureCompletesExceptionally() {
+        final ThrottledFuture<?> future = ThrottledFuture.of(() -> {
+            throw new RuntimeException();
+        });
+        future.run(null);
+        assertTrue(future.isCompletedExceptionally());
+    }
+
+    @Test
+    void isCompletedExceptionallyReturnsFalseWhenFutureCompletesNormally() {
+        final ThrottledFuture<?> future = ThrottledFuture.of(() -> null);
+        future.run(null);
+        assertFalse(future.isCompletedExceptionally());
+        assertFalse(future.isCancelled());
+        assertTrue(future.isDone());
+    }
+
+    @Test
+    void isCompletedExceptionallyReturnsTrueWhenFutureIsCancelled() {
+        final ThrottledFuture<?> future = ThrottledFuture.of(() -> null);
+        future.cancel(false);
+        assertTrue(future.isCompletedExceptionally());
+        assertTrue(future.isCancelled());
+        assertTrue(future.isDone());
+    }
+
+    @Test
+    void thenActionIsExecutedWhenFutureIsAlreadyCompletedExceptionally() {
         final AtomicBoolean completed = new AtomicBoolean(false);
         final ThrottledFuture<?> future = ThrottledFuture.of(() -> {
             throw new RuntimeException();

--- a/src/test/java/cz/cvut/kbss/termit/util/throttle/ThrottledFutureTest.java
+++ b/src/test/java/cz/cvut/kbss/termit/util/throttle/ThrottledFutureTest.java
@@ -299,8 +299,8 @@ class ThrottledFutureTest {
 
     @Test
     void transferUpdatesSecondFutureWithCallbacks() {
-        final Consumer<String> firstCallback = (result) -> {};
-        final Consumer<String> secondCallback = (result) -> {};
+        final Consumer<ThrottledFuture<String>> firstCallback = (result) -> {};
+        final Consumer<ThrottledFuture<String>> secondCallback = (result) -> {};
         final ThrottledFuture<String> firstFuture = ThrottledFuture.of(()->"").then(firstCallback);
         final ThrottledFuture<String> secondFuture = ThrottledFuture.of(()->"").then(secondCallback);
         final ThrottledFuture<String> mocked = mock(ThrottledFuture.class);
@@ -323,14 +323,14 @@ class ThrottledFutureTest {
 
     @Test
     void callbacksAreClearedAfterTransferring() {
-        final Consumer<String> firstCallback = (result) -> {};
-        final Consumer<String> secondCallback = (result) -> {};
+        final Consumer<ThrottledFuture<String>> firstCallback = (result) -> {};
+        final Consumer<ThrottledFuture<String>> secondCallback = (result) -> {};
         final ThrottledFuture<String> future = ThrottledFuture.of(()->"").then(firstCallback).then(secondCallback);
         final ThrottledFuture<String> mocked = mock(ThrottledFuture.class);
 
         future.transfer(mocked);
 
-        final ArgumentCaptor<List<Consumer<String>>> captor = ArgumentCaptor.forClass(List.class);
+        final ArgumentCaptor<List<Consumer<ThrottledFuture<String>>>> captor = ArgumentCaptor.forClass(List.class);
 
         verify(mocked).update(notNull(), captor.capture());
         // captor takes the original list from the future
@@ -386,8 +386,8 @@ class ThrottledFutureTest {
 
     @Test
     void updateAddsCallbacksToTheCurrentOnes() {
-        final Consumer<String> callback = result -> {};
-        final Consumer<String> originalCallback = result -> {};
+        final Consumer<ThrottledFuture<String>> callback = result -> {};
+        final Consumer<ThrottledFuture<String>> originalCallback = result -> {};
         final ThrottledFuture<String> future = ThrottledFuture.of(() -> "").then(originalCallback);
 
         future.update(()->"", List.of(callback));

--- a/src/test/java/cz/cvut/kbss/termit/util/throttle/ThrottledFutureTest.java
+++ b/src/test/java/cz/cvut/kbss/termit/util/throttle/ThrottledFutureTest.java
@@ -85,8 +85,7 @@ class ThrottledFutureTest {
 
     @Test
     void thenActionIsExecutedSynchronouslyWhenFutureIsAlreadyDoneAndNotCanceled() {
-        final Object result = new Object();
-        final ThrottledFuture<?> future = ThrottledFuture.of(() -> result);
+        final ThrottledFuture<?> future = ThrottledFuture.of(() -> null);
         final AtomicBoolean completed = new AtomicBoolean(false);
         final AtomicReference<Object> futureResult = new AtomicReference<>(null);
         future.run(null);
@@ -97,25 +96,24 @@ class ThrottledFutureTest {
             futureResult.set(fResult);
         });
         assertTrue(completed.get());
-        assertEquals(result, futureResult.get());
+        assertEquals(future, futureResult.get());
     }
 
     @Test
-    void thenActionIsNotExecutedWhenFutureIsAlreadyCancelled() {
+    void thenActionIsExecutedWhenFutureIsAlreadyCancelled() {
         final ThrottledFuture<?> future = ThrottledFuture.of(Object::new);
         final AtomicBoolean completed = new AtomicBoolean(false);
         future.cancel(false);
         assertTrue(future.isCancelled());
         future.then(result -> completed.set(true));
-        assertFalse(completed.get());
+        assertTrue(completed.get());
     }
 
     @Test
     void thenActionIsExecutedOnceFutureIsRun() {
-        final Object result = new Object();
         final AtomicBoolean completed = new AtomicBoolean(false);
         final AtomicReference<Object> fResult = new AtomicReference<>(null);
-        final ThrottledFuture<?> future = ThrottledFuture.of(() -> result);
+        final ThrottledFuture<?> future = ThrottledFuture.of(() -> null);
         future.then(futureResult -> {
             completed.set(true);
             fResult.set(futureResult);
@@ -124,22 +122,22 @@ class ThrottledFutureTest {
         assertFalse(completed.get()); // action was not executed yet
         future.run(null);
         assertTrue(completed.get());
-        assertEquals(result, fResult.get());
+        assertEquals(future, fResult.get());
     }
 
     @Test
-    void thenActionIsNotExecutedOnceFutureIsCancelled() {
+    void thenActionIsExecutedOnceFutureIsCancelled() {
         final Object result = new Object();
         final AtomicBoolean completed = new AtomicBoolean(false);
         final ThrottledFuture<?> future = ThrottledFuture.of(() -> result);
         future.then(futureResult -> completed.set(true));
         assertFalse(completed.get()); // action was not executed yet
         future.cancel(false);
-        assertFalse(completed.get());
+        assertTrue(completed.get());
     }
 
     @Test
-    void thenActionIsNotExecutedWhenFutureCompletedExceptionally() {
+    void thenActionIsExecutedWhenFutureCompletedExceptionally() {
         final AtomicBoolean completed = new AtomicBoolean(false);
         final ThrottledFuture<?> future = ThrottledFuture.of(() -> {
             throw new RuntimeException();
@@ -147,7 +145,7 @@ class ThrottledFutureTest {
         future.run(null);
         assertFalse(completed.get());
         future.then(futureResult -> completed.set(true));
-        assertFalse(completed.get());
+        assertTrue(completed.get());
     }
 
     @Test

